pkg/monitor/lease.go

```diff
@@ -0,0 +1,137 @@
+package monitor
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/milosgajdos/tenus"
+	"github.com/openshift/baremetal-runtimecfg/pkg/config"
+	"github.com/openshift/baremetal-runtimecfg/pkg/utils"
+	"github.com/sirupsen/logrus"
+)
+
+/*
+	After reading the cmd flags and before executing the keepalived-monitor:
+	1. Get the directory of args[2] = path_to_config (e.g. /etc/keepalived/keepalived.conf) [V]
+	2. Check if the there's a file named "monitor.conf" in that directory. If true, then: [V]
+		2.1. for [API interface, Ingress interface]: . [V]
+			2.1.1 Generate a unique MAC address
+				2.1.1.1. The first 24 bits of the MAC should be the same as the real interface mac address.
+						Get the interface by using `GetInterfaceAndNonVIPAddr()` and after that get its mac address.
+				2.1.1.2. The last 24 bits are generated using some hash function.
+						An example is the way the VRIDs are generated in `PopulateVRIDs()`
+			2.1.2. Create an interface type macvlan per vip using the generated mac address [V]
+			2.1.3. Run dhclient with the new vip interface in the background (goroutine)
+	3. Continue regular run
+*/
+
+const monitorConf = "monitor.conf"
+
+type VIP struct {
+	name string
+	ip   net.IP
+}
+
+func needLease(cfgPath string) (bool, error) {
+	monitorConfFile := filepath.Join(filepath.Dir(cfgPath), monitorConf)
+
+	_, err := os.Stat(monitorConfFile)
+
+	if err == nil {
+		return true, nil
+	}
+	if os.IsNotExist(err) {
+		return false, nil
+	}
+	return false, err
+}
+
+func leaseVIPs(vips []VIP) error {
+	for _, vip := range vips {
+		if err := leaseVIP(vip.name, vip.ip); err != nil {
+			log.WithFields(logrus.Fields{
+				"vip": vip,
+			}).Error("Failed to lease a vip")
+			return err
+		}
+	}
+
+	return nil
+}
+
+func leaseVIP(name string, vip net.IP) error {
+	iface, err := createVIPInterface(name, vip)
+
+	if err != nil {
+		return err
+	}
+
+	// Renew with dhclient. Maybe can be replaced with https://github.com/insomniacslk/dhcp
+	// TODO: define timeout
+	cmd := exec.Command("dhclient", "-d", "-timeout", "5", iface.Name)
```

Comment:
This seems problematic. -d is going to keep it in the foreground, which means either this never returns or it returns once and never renews the lease (it's not clear to me which). Since the leaseVIPs call isn't in the monitor loop, I assume the intent was to have dhclient remain running in the background? Otherwise the lease call needs to be moved into the loop. Edit: I see from the comment above that the intent was for this to be run in a goroutine. That isn't happening in the current version so this still needs changes.

Also, we can't actually configure the address on the node since we want keepalived to manage the assignment of the VIP. I believe that means we need -n for this command.

Reply:
Don't we also need -n? We don't want dhclient configuring the VIP on the interface or it may conflict with what keepalived does with it.

Reply:
IMHO Not configuring the interface is handled by the script passed to "-sf"

"-n" does not really work

Reply:
I think that for now we can just pass `/bin/true`